man/pward.1
.\" groff -man -Tascii pward.1
.\" groff -man -Tascii pward.1 | col -b
.TH PWARD 1 "July 2010" Linux "User Manuals"
.SH NAME
pward \- monitor the lifecycle of a set processes.
.SH SYNOPSIS
.B pward
[
.I OPTION
] ... [
.I
PID
] ...
.SH DESCRIPTION
.B pward
monitors a set of processes given at the command-line. By default,
pward stops when all specified processes have died.
Optionnally, it stops when a specified number of processes has stopped,
or when less than a specified number of processes is still running.
It can be configured to execute a specified command before stopping.
.SH OPTIONS
.TP
\fB\-r\fR, \fB\-\-running\fR=n
stop monitoring when no more than n processes are still running (default 0)
.TP
\fB\-s\fR, \fB\-\-stopped\fR=n
stop monitoring when at least n processes have stopped (default 1)
.TP
\fB\-e\fR, \fB\-\-exec\fR=\fIcommand\fR
Execute \fIcommand\fR after the treshold is met, and before pward is stopped.
The command  is executed in the starting directory.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
summarize monitored processes at startup.
.TP
\fB\-b\fR, \fB\-\-batch\fR
don't perform sanity checks at input. This can be useful when invoked from a script
.TP
\fB\-h\fR, \fB\-\-help\fR
print a help message at startup
.TP
\fB[PID]\fR
space separated list of process-ids to be monitored. 

.SH FILES

.SH ENVIRONMENT

.SH DIAGNOSTICS
The following diagnostics may be issued on stderr:

non numeric parameter
.RS
The input parameter is not a number
.RE

command failed
.RS
the command in the -exec option failed to execute
.RE

.SH BUGS
pward checks processes by polling the proc-table.
This is portable but an implementation with modern kernel features would be more efficient.

.SH EXAMPLES
.B pward `pgrep emerge` -v -i10 -e"shutdown -h +10"

Shuts down the machine 10 minutes after emerge finished.
Some info about the monitored processes is dumped to the console at startup.
The polling interval is set to 10 seconds.
The user needs the rights to shutdown the machine for this to work.

.SH AUTHOR
Dirk Moermans <dirkmoermans @ gmail.com>
.SH "SEE ALSO"
.BR pgrep (1),
.BR pkill (1),
.BR top (1)


src/pward.c
#define _GNU_SOURCE
#include "proc_impl.h"

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

static void
print_usage(const char* name)
{
   printf(
      PACKAGE": [OPTION]... [PID]... \n"
      "  -r, --running=n: stop if no more than n processes are still running (default 0)\n"
      "  -s, --stopped=n: stop if at least n processes have stopped (default 1)\n"
      "  -e, --exec=\"cmd\": command to execute when treshold is met\n"
      "  -i, --interval=n: polling interval in seconds (default 1)\n"
      "  -v, --verbose: summarize monitored processes at startup\n"
      "  -b, --batch: no sanity checks on input (useful when scripting)\n"
      "  -h, --help: print this help message\n\n"
      "  [PID]: space separated list of process-ids to be monitored.\n\n"
      "default behaviour: run until all specified processes have died\n"
      "pward-version: "PACKAGE_VERSION" \n");
}

static inline size_t checked_strtoul(const char* arg, bool* success)
{
   errno=0;
   char* endptr;
   long signed_result=strtol(arg,&endptr,10);
   *success= (errno != ERANGE) &&(errno != EINVAL)
      && (endptr!=arg) && (*endptr=='\0')
      &&(signed_result>0);
   return (size_t)signed_result;
}

int main(int argc,const char* argv[])
{
   bool verbose=false;
   bool batch=false;
   const char* cmd="";
   size_t running=0;

   bool stopCondition=false;
   size_t stopped=1;
   int nInterval=1; /* seconds */

   while(1)
   {
      static const struct option long_options[]=
         {
            {"verbose", 0, 0, 'v'},
            {"help", 0, 0, 'h'},
            {"running", 2, 0, 'r'},
            {"stopped", 2, 0, 's'},
            {"exec", 1, 0, 'e'},
            {"batch", 0, 0, 'b'},
            {"interval", 1, 0, 'i'},
            {0, 0, 0, 0}
         };

      int c = getopt_long(argc,(char* const*)argv, "vhr::s::e:bi:",
			  long_options,NULL);
      if(c==-1)
         break;

      switch(c)
      {
         case 'v':
            verbose=1;
            break;
         case 'h':
            print_usage(argv[0]);
            return -1;
         case 'r':
            if(optarg!=NULL)
	    {
               bool success;
               running=checked_strtoul(optarg,&success);
               if(!success)
               {
		  fprintf(stderr,"invalid parameter: '%s' "
			  "as number of running processes\n",optarg);
		  return -1;
               }
	    }
            break;
         case 's':
            stopCondition=1;
            if(optarg!=NULL)
	    {
               bool success;
               stopped=checked_strtoul(optarg,&success);
               if(!success)
               {
		  fprintf(stderr,"invalid parameter: '%s' "
			  "as number as stopped processes\n",optarg);
		  return -1;
               }
	    }
            break;
         case 'e':
            cmd=strdupa(optarg);
            break;
         case 'b':
            batch=1;
            break;
         case 'i':
         {
	    bool success;
	    stopped=checked_strtoul(optarg,&success);
	    if(!success)
            {
               fprintf(stderr,"invalid parameter: '%s' "
                       "as interval time\n",optarg);
               return -1;
            }
	    break;
         }
         case '?':
            return -1; /* missing opt argument */
      }
   }

   int nLastOptionIndex=optind;
   size_t nProcsInit=argc-nLastOptionIndex;

   if(!nProcsInit && !batch)
   {
      fprintf(stderr,"WARNING: no processes are monitored\n\n");
   }
   /* recalculate 'stop' in terms of number of running processes */
   if(stopCondition)
   {
      if(nProcsInit<stopped)
      {
         if(!batch)
	    fprintf(stderr,"WARNING: requesting more stopped processes (%zu) "
                    "than the number of processes at startup (%zu)\n",stopped,nProcsInit);
      }
      else if(running<nProcsInit-stopped)
         running=nProcsInit-stopped;
   }

   pid_t pids[nProcsInit];
   for(int i=0;i<nProcsInit;++i)
   {
      bool success;
      pids[i]=checked_strtoul(argv[nLastOptionIndex+i],&success);
      if(!success)
      {
         fprintf(stderr,"invalid parameter: '%s' "
                 "as process id\n",argv[nLastOptionIndex+i]);
         return -1;
      }
   }

   int result=proc_observe_processes
      (nProcsInit,pids,running,batch,verbose,nInterval);
   if(result)
      return result;

   if(verbose)
      printf("condition met: %zu processes left\n", running);

   if(*cmd!='\x0')
   {
      if(verbose)
         printf("executing command: '%s'\n",cmd);
      if(!system(cmd))
         fprintf(stderr,"command '%s' failed to spawn\n",cmd);
   }
   return 0;
}
